File: src/Farkle/Builder/Terminals.fs
// Copyright (c) 2020 Theodore Tsirpanis
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT

/// <summary>Some designtime Farkles that
/// are commonly used in many grammars.</summary>
/// <remarks>
/// These functions take a name and create
/// a designtime Farkle which is meant to be
/// reused everywhere it is needed in the grammar.
/// Creating and using many designtime Farkles
/// of the same or similar kind
/// will almost certainly lead to an error.</remarks>
[<RequireQualifiedAccess>]
module Farkle.Builder.Terminals

open System.ComponentModel
open Farkle.Builder
open Farkle.Builder.Regex
open System
open System.Globalization
open System.Text

/// This module is public due to compiler limitations.
/// Do not use it; it is subject to be removed or altered at any time.
// It's public because of https://github.com/dotnet/fsharp/issues/8348
[<EditorBrowsable(EditorBrowsableState.Never)>]
module Internal =
    #if !MODERN_FRAMEWORK
    type Characters = string
    let inline toCharacters (x: ReadOnlySpan<char>): Characters = x.ToString()
    #else
    type Characters = ReadOnlySpan<char>
    let inline toCharacters x: Characters = x
    #endif

open Internal

/// A `Regex` that recognizes a series of decimal digits.
/// Leading zeros are accepted.
// Seriously though, why not? Looking at you, JSON!
[<CompiledName("UnsignedIntRegex")>]
let unsignedIntRegex = Number |> chars |> atLeast 1

/// Creates a designtimeFarkle that parses an unsigned decimal integer
/// into the desired number type. No bounds checking is performed.
/// Using this function from a language other than F# will throw an exception.
[<CompiledName("FSharpOnlyGenericUnsigned"); RequiresExplicitTypeArguments; NoDynamicInvocation>]
let inline genericUnsigned< ^TInt when ^TInt: (static member Parse: Characters * NumberStyles * IFormatProvider -> ^TInt)> name =
    terminal name
        (T (fun _ x ->
            try
                (^TInt: (static member Parse: Characters * NumberStyles * IFormatProvider -> ^TInt)
                    (toCharacters x, NumberStyles.None, NumberFormatInfo.InvariantInfo))
            with
            | e -> error e.Message))
        unsignedIntRegex

/// A `Regex` that recognizes a series of decimal digits
/// that might be prefixed with a minus sign "-".
/// Leading zeros are accepted.
[<CompiledName("SignedIntRegex")>]
let signedIntRegex = (optional <| char '-') <&> unsignedIntRegex

/// Creates a designtime Farkle that parses a signed decimal integer
/// into the desired number type. No bounds checking is performed.
/// The type parameter must support the unary negation operator.
/// Using this function from a language other than F# will throw an exception.
[<CompiledName("FSharpOnlyGenericSigned"); RequiresExplicitTypeArguments; NoDynamicInvocation>]
let inline genericSigned< ^TInt when ^TInt: (static member Parse: Characters * NumberStyles * IFormatProvider -> ^TInt) and ^TInt : (static member (~-): ^TInt -> ^TInt)> name =
    terminal name
        (T (fun _ x ->
            try
                (^TInt: (static member Parse: Characters * NumberStyles * IFormatProvider -> ^TInt)
                    (toCharacters x, NumberStyles.Integer, NumberFormatInfo.InvariantInfo))
            with
            | e -> error e.Message))
        signedIntRegex

/// Creates a designtime Farkle that parses and returns a
/// signed 32-bit signed integer. No bounds checking is performed.
[<CompiledName("Int32")>]
let int name = genericSigned<int> name

/// Creates a designtime Farkle that parses and returns a
/// signed 64-bit signed integer. No bounds checking is performed.
[<CompiledName("Int64")>]
let int64 name = genericSigned<int64> name

/// Creates a designtime Farkle that parses and returns a
/// signed 32-bit unsigned integer. No bounds checking is performed.
[<CompiledName("UInt32")>]
let uint32 name = genericUnsigned<uint32> name

/// Creates a designtime Farkle that parses and returns a
/// signed 64-bit unsigned integer. No bounds checking is performed.
[<CompiledName("UInt64")>]
let uint64 name = genericUnsigned<uint64> name

/// A `Regex` that recognizes an unsigned real number.
/// The number is expected to be written in scientific
/// notation, with the decimal point symbol being the dot.
/// Numbers before or after the dot are optional, but they
/// must exist in at least one of the two places.
[<CompiledName("UnsignedRealRegex")>]
let unsignedRealRegex =
    let numberStar = chars Number |> star
    let atLeastOneNumber = chars Number <&> numberStar
    let dotOptional = char '.' |> optional
    concat [
        choice [
            // There has to be at least one digit
            // either before or after the dot.
            // Or no dot at all!
            atLeastOneNumber <&> dotOptional <&> numberStar
            numberStar <&> dotOptional <&> atLeastOneNumber
        ]
        [chars "eE"; optional <| chars "+-"; atLeastOneNumber]
        |> concat
        |> optional
    ]

/// Like `unsignedRealRegex`, but with an optional
/// sign in the beginning being allowed.
[<CompiledName("SignedRealRegex")>]
let signedRealRegex = (optional <| char '-') <&> unsignedRealRegex

/// Creates a designtime Farkle that parses a real number
/// into the desired number type. No bounds checking is performed.
/// Using this function from a language other than F# will throw an exception.
[<CompiledName("FSharpOnlyGenericReal"); RequiresExplicitTypeArguments; NoDynamicInvocation>]
let inline genericReal< ^TReal when ^TReal: (static member Parse: Characters * NumberStyles * IFormatProvider -> ^TReal)> allowSign name =
    let regex =
        if allowSign then
            signedRealRegex
        else
            unsignedRealRegex
    terminal name
        (T (fun _ x ->
            try
                (^TReal: (static member Parse: Characters * NumberStyles * IFormatProvider -> ^TReal)
                    (toCharacters x, NumberStyles.Float, NumberFormatInfo.InvariantInfo))
            with
            | e -> error e.Message))
        regex

/// Creates a designtime Farkle that parses and returns
/// a signed single-precision floating-point number. The number
/// is expected to be written in scientific notation, with
/// the decimal point symbol being the dot. Special values
/// such as NaN or infinity are not recognized.
[<CompiledName("Single")>]
let float32 name = genericReal<float32> true name

/// Creates a designtime Farkle that parses and returns
/// a signed double-precision floating-point number. The number
/// is expected to be written in scientific notation, with
/// the decimal point symbol being the dot. Special values
/// such as NaN or infinity are not recognized.
[<CompiledName("Double")>]
let float name = genericReal<float> true name

/// Creates a designtime Farkle that parses and returns
/// a signed decimal floating-point number. The number
/// is expected to be written in scientific notation, with
/// the decimal point symbol being the dot.
[<CompiledName("Decimal")>]
let decimal name = genericReal<decimal> true name

let private stringTransformer = T (fun _ span ->
    let mutable span = span.Slice(1, span.Length - 2)
    let sb = StringBuilder(span.Length)

    #if !MODERN_FRAMEWORK
    let mutable i = 0
    while i < span.Length do
        let c = span.[i]
        i <- i + 1
        match c with
        | '\\' ->
            let c = span.[i]
            i <- i + 1
            match c with
            | 'a' -> sb.Append '\a'
            | 'b' -> sb.Append '\b'
            | 'f' -> sb.Append '\f'
            | 'n' -> sb.Append '\n'
            | 'r' -> sb.Append '\r'
            | 't' -> sb.Append '\t'
            | 'u' ->
                let hexCode =
                    UInt16.Parse(span.Slice(i, 4).ToString(), NumberStyles.HexNumber)
                i <- i + 4
                sb.Append(Operators.char hexCode)
            | 'v' -> sb.Append '\v'
            | c -> sb.Append c
        | c -> sb.Append c
        |> ignore
    #else
    while not span.IsEmpty do
        match span.IndexOf('\\') with
        | -1 ->
            sb.Append(span) |> ignore
            span <- ReadOnlySpan.Empty
        | backslashIdx ->
            let charactersBeforeBackslash = span.Slice(0, backslashIdx)
            sb.Append(charactersBeforeBackslash) |> ignore
            let escapePayloadLength =
                match span.[backslashIdx + 1] with
                | 'a' -> sb.Append '\a' |> ignore; 1
                | 'b' -> sb.Append '\b' |> ignore; 1
                | 'f' -> sb.Append '\f' |> ignore; 1
                | 'n' -> sb.Append '\n' |> ignore; 1
                | 'r' -> sb.Append '\r' |> ignore; 1
                | 't' -> sb.Append '\t' |> ignore; 1
                | 'u' ->
                    let hexCode =
                        UInt16.Parse(span.Slice(backslashIdx + 2, 4), NumberStyles.HexNumber)
                    sb.Append(Operators.char hexCode) |> ignore
                    5
                | 'v' -> sb.Append '\v' |> ignore; 1
                | c -> sb.Append c |> ignore; 1
            span <- span.Slice(backslashIdx + 1 + escapePayloadLength)
    #endif

    sb.ToString())

/// <summary>Creates a designtime Farkle that can recognize a C-like
/// string. It supports escaping with a backslash "\\".</summary>
/// <param name="escapeChars">A sequence of valid escape characters after the backslash.
/// If one of <c>abfnrtv</c> is included, it will be treated as a developer would expect.
/// Any other character will be output verbatim. The backslash and the string delimiter
/// are always escaped. Specifying <c>u</c> will be ignored.</param>
/// <param name="allowEscapeUnicode">Whether to allow escaping Unicode characters with
/// <c>\\u</c>, followed by exactly four case-insensitive hex digits.</param>
/// <param name="multiline">Whether to allow line breaks in the string. They will
/// be interpreted as literal line breaks.</param>
/// <param name="delim">The character that marks the start and the end of the string.
/// In most programming languages, it is either a single or a double quote.</param>
/// <param name="name">The name of the resulting designtime Farkle.</param>
[<CompiledName("StringEx")>]
let stringEx escapeChars allowEscapeUnicode multiline delim name =
    let escapeChars =
        if allowEscapeUnicode then
            Seq.filter ((<>) 'u') escapeChars
        else
            escapeChars
    let regex =
        let stringCharacters =
            [delim; '\\'] @ if multiline then ['\r'; '\n'] else []
            |> allButChars
        concat [
            char delim
            star <| choice [
                stringCharacters
                concat [
                    char '\\'
                    choice [
                        char '\\'
                        char delim
                        chars escapeChars
                        if allowEscapeUnicode then
                            char 'u' <&> (repeat 4 <| chars "1234567890ABCDEFabcdef")
                    ]
                ]
            ]
            char delim
        ]
    terminal name stringTransformer regex

/// Creates a designtime Farkle that can recognize a single-line, C-like string.
[<CompiledName("String")>]
let string delim name = stringEx "abfnrtv" true false delim name

File: src/Farkle/Builder/Terminals.cs
// Copyright © Theodore Tsirpanis and Contributors.
// SPDX-License-Identifier: MIT

using Farkle.Parser;
using System.Collections.Immutable;
using System.Globalization;
using System.Numerics;
using System.Text;

namespace Farkle.Builder;

/// <summary>
/// Provides factory methods for common terminal symbols.
/// </summary>
public static class Terminals
{
    private static readonly Regex s_optionalMinus = Regex.OneOf(('-', '-')).Optional().CaseSensitive();

    private static readonly Regex s_unsignedIntRegex = Regex.OneOf(('0', '9')).AtLeast(1).CaseSensitive();

    private static readonly Regex s_intRegex = s_optionalMinus + s_unsignedIntRegex;

    private static readonly Regex s_unsignedFloatRegex = Regex.Join(
        Regex.OneOf('-').Optional(),
        s_unsignedIntRegex,
        Regex.OneOf('.'),
        s_unsignedIntRegex,
        Regex.Join(
            Regex.OneOf('e', 'E'),
            Regex.OneOf('+', '-').Optional(),
            s_unsignedIntRegex
        ).Optional()
    ).CaseSensitive();

    private static readonly Regex s_signedFloatRegex = s_optionalMinus + s_unsignedFloatRegex;

    private static string TransformString(ref ParserState state, ReadOnlySpan<char> str)
    {
        str = str[1..^1];

        // Fast path if there are no escape sequences.
        if (!str.Contains('\\'))
        {
            return str.ToString();
        }

        var sb = new StringBuilder(str.Length);
        while (!str.IsEmpty)
        {
            switch (str.IndexOf('\\'))
            {
                case -1:
                    sb.Append(str);
                    str = [];
                    break;
                case int backslashIdx:
                    sb.Append(str[..backslashIdx]);
                    char charAtBackslash = str[backslashIdx + 1];
                    char c = charAtBackslash switch
                    {
                        'a' => '\a',
                        'b' => '\b',
                        'e' => '\x1B', // Escape character
                        'f' => '\f',
                        'n' => '\n',
                        'r' => '\r',
                        't' => '\t',
                        'u' => (char)ushort.Parse(str.Slice(backslashIdx + 2, 4), NumberStyles.HexNumber),
                        'v' => '\v',
                        _ => charAtBackslash
                    };
                    sb.Append(c);
                    str = str[(backslashIdx + (charAtBackslash == 'u' ? 6 : 2))..];
                    break;
            }
        }
        return sb.ToString();
    }

    private static T TransformInteger<T>(ref ParserState state, ReadOnlySpan<char> str) where T : INumberBase<T>
    {
        try
        {
            return T.Parse(str, NumberStyles.Integer, CultureInfo.InvariantCulture);
        }
        catch (Exception ex)
        {
            throw new ParserApplicationException(ex.Message);
        }
    }

    private static T TransformFloat<T>(ref ParserState state, ReadOnlySpan<char> str) where T : INumberBase<T>
    {
        try
        {
            return T.Parse(str, NumberStyles.Float, CultureInfo.InvariantCulture);
        }
        catch (Exception ex)
        {
            throw new ParserApplicationException(ex.Message);
        }
    }

    /// <summary>
    /// Creates a terminal that matches an unsigned integer.
    /// </summary>
    /// <typeparam name="T">The type this terminal returns. Must implement <see cref="INumberBase{TSelf}"/>.</typeparam>
    /// <param name="name">The terminal's name.</param>
    /// <remarks>
    /// This terminal matches non-empty sequences of decimal digits. Leading zeroes are not prohibited.
    /// </remarks>
    public static IGrammarSymbol<T> UnsignedInteger<T>(string name) where T : INumberBase<T> =>
        Terminal.Create(name, s_unsignedIntRegex, TransformInteger<T>);

    /// <summary>
    /// Creates a terminal that matches a signed integer.
    /// </summary>
    /// <typeparam name="T">The type this terminal returns. Must implement <see cref="ISignedNumber{TSelf}"/>.</typeparam>
    /// <param name="name">The terminal's name.</param>
    /// <remarks>
    /// This terminal matches non-empty sequences of decimal digits. Leading zeroes are not prohibited.
    /// A leading minus sign is also allowed.
    /// </remarks>
    public static IGrammarSymbol<T> SignedInteger<T>(string name) where T : ISignedNumber<T> =>
        Terminal.Create(name, s_intRegex, TransformInteger<T>);

    /// <summary>
    /// Creates a terminal that matches an unsigned floating-point number.
    /// </summary>
    /// <typeparam name="T">The type this terminal returns. Must implement <see cref="IFloatingPoint{TSelf}"/>.</typeparam>
    /// <param name="name">The terminal's name.</param>
    /// <remarks>
    /// <para>
    /// This terminal matches non-empty sequences of decimal digits, separated by a dot, and optionally followed by an
    /// exponent of the form <c>[eE][+-]?[0-9]+</c>.
    /// </para>
    /// <para>
    /// The set of accepted strings has changed in Farkle 7. Previously the dot, and digits either only
    /// before or only after it, were optional, which made the terminal match integer literals, but caused
    /// conflicts. Making them mandatory avoids these conflicts and improves simplicity and predictability.
    /// </para>
    /// </remarks>
    public static IGrammarSymbol<T> UnsignedFloat<T>(string name) where T : IFloatingPoint<T> =>
        Terminal.Create(name, s_unsignedFloatRegex, TransformFloat<T>);

    /// <summary>
    /// Creates a terminal that matches a signed floating-point number.
    /// </summary>
    /// <typeparam name="T">The type this terminal returns. Must implement <see cref="IFloatingPoint{TSelf}"/>.</typeparam>
    /// <param name="name">The terminal's name.</param>
    /// <remarks>
    /// <para>
    /// This terminal matches non-empty sequences of decimal digits, separated by a dot, and optionally followed by an
    /// exponent of the form <c>[eE][+-]?[0-9]+</c>. A leading minus sign is also allowed.
    /// </para>
    /// <para>
    /// The set of accepted strings has changed in Farkle 7. Previously the dot, and digits either only
    /// before or only after it, were optional, which made the terminal match integer literals, but caused
    /// conflicts. Making them mandatory avoids these conflicts and improves simplicity and predictability.
    /// </para>
    /// </remarks>
    public static IGrammarSymbol<T> SignedFloat<T>(string name) where T : IFloatingPoint<T> =>
        Terminal.Create(name, s_signedFloatRegex, TransformFloat<T>);

    /// <inheritdoc cref="SignedInteger{T}(string)"/>
    public static IGrammarSymbol<int> Int32(string name) => SignedInteger<int>(name);

    /// <inheritdoc cref="SignedInteger{T}(string)"/>
    public static IGrammarSymbol<long> Int64(string name) => SignedInteger<long>(name);

    /// <inheritdoc cref="UnsignedInteger{T}(string)"/>
    public static IGrammarSymbol<uint> UInt32(string name) => UnsignedInteger<uint>(name);

    /// <inheritdoc cref="UnsignedInteger{T}(string)"/>
    public static IGrammarSymbol<ulong> UInt64(string name) => UnsignedInteger<ulong>(name);

    /// <inheritdoc cref="SignedFloat{T}(string)"/>
    public static IGrammarSymbol<float> Single(string name) => SignedFloat<float>(name);

    /// <inheritdoc cref="SignedFloat{T}(string)"/>
    public static IGrammarSymbol<double> Double(string name) => SignedFloat<double>(name);

    /// <inheritdoc cref="SignedFloat{T}(string)"/>
    public static IGrammarSymbol<decimal> Decimal(string name) => SignedFloat<decimal>(name);

    /// <summary>
    /// Creates a terminal that matches a C-style string.
    /// </summary>
    /// <param name="name">The terminal's name.</param>
    /// <param name="delimiter">The character that starts and ends the string.</param>
    /// <param name="escapeChars">The characters that are allowed to be escaped with a <c>\</c>.
    /// <c>\</c> and <paramref name="delimiter"/> are always allowed to be escaped.</param>
    /// <param name="multiLine">Whether new line characters are allowed.</param>
    /// <remarks>
    /// The following escape sequences if enabled in <paramref name="escapeChars"/> have special meaning:
    /// <list type="bullet">
    /// <item>
    /// <c>\a</c>, <c>\b</c>, <c>\f</c>, <c>\n</c>, <c>\r</c>, <c>\t</c> and <c>\v</c> correspond to the
    /// same characters as in C.
    /// </item>
    /// <item>
    /// <c>\e</c> corresponds to the <c>U+001B</c> escape character.
    /// </item>
    /// <item>
    /// <c>\UXXXX</c>, where <c>X</c> is a case-insensitive hexadecimal number, corresponds
    /// to the <c>U+XXXX</c> character.
    /// </item>
    /// </list>
    /// Escape sequences with all other characters are treated as the character itself.
    /// </remarks>
    public static IGrammarSymbol<string> String(string name, char delimiter, string escapeChars, bool multiLine)
    {
        ArgumentNullException.ThrowIfNull(name);
        ArgumentNullException.ThrowIfNull(escapeChars);

        var regexDelimiter = Regex.Literal(delimiter);
        var regexBackslash = Regex.Literal('\\');
        var regex =
            Regex.Join(
                regexDelimiter,
                Regex.Choice(
                    Regex.NotOneOf(multiLine ? ['\n', '\r', '\\', delimiter] : ['\\', delimiter]),
                    Regex.Join(
                        regexBackslash,
                        Regex.Choice(
                            regexBackslash,
                            regexDelimiter,
                            Regex.OneOf(escapeChars.Where(c => c != 'u')),
                            escapeChars.Contains('u')
                                ? Regex.Join(
                                    Regex.Literal('u'),
                                    Regex.OneOf(('0', '9'), ('a', 'f'), ('A', 'F')).Repeat(4)
                                )
                                : Regex.Void
                        )
                    )
                ).ZeroOrMore(),
                regexDelimiter
            ).CaseSensitive();

        return Terminal.Create(name, regex, TransformString);
    }

    /// <summary>
    /// Creates a terminal that matches a single-line C-style string.
    /// </summary>
    /// <param name="name">The terminal's name.</param>
    /// <param name="delimiter">The character that starts and ends the string.</param>
    /// <remarks>
    /// All escape sequences described in <see cref="String(string, char, string, bool)"/>
    /// are supported.
    /// </remarks>
    public static IGrammarSymbol<string> String(string name, char delimiter) =>
        String(name, delimiter, "abefnrtvu", false);
}
